include/lbann/utils/cuda.hpp

```diff
@@ -67,5 +69,69 @@
 #define CHECK_CUDA(cuda_call) (cuda_call)
 #endif // #ifdef LBANN_DEBUG
 
+namespace lbann {
+namespace cuda {
+
+namespace thrust {
+
+/** GPU memory allocator that can interact with Thrust.
+ *  Uses Hydrogen's CUB memory pool if available.
+ */
+template <typename T = El::byte>
+class allocator
+  : public ::thrust::detail::tagged_allocator<
+      T,
+      ::thrust::system::cuda::tag,
+      ::thrust::pointer<T, ::thrust::system::cuda::tag>> {
+private:
+  typedef typename ::thrust::detail::tagged_allocator<
+    T,
+    ::thrust::system::cuda::tag,
+    ::thrust::pointer<T, ::thrust::system::cuda::tag>> parent_class;
+
+  /** Active CUDA stream. */
+  cudaStream_t m_stream;
+
+public:
+  typedef typename parent_class::value_type value_type;
+  typedef typename parent_class::pointer    pointer;
+  typedef typename parent_class::size_type  size_type;
+
+  allocator(cudaStream_t stream = El::GPUManager::Stream())
+    : m_stream(stream) {}
+
+  /** Allocate GPU buffer. */
+  pointer allocate(size_type size) {
+    value_type* buffer = nullptr;
+#ifdef HYDROGEN_HAVE_CUB
+    auto& memory_pool = El::cub::MemoryPool();
+    CHECK_CUDA(memory_pool.DeviceAllocate(reinterpret_cast<void**>(&buffer),
+                                          size * sizeof(value_type),
+                                          m_stream));
+#else
+    CHECK_CUDA(cudaMalloc(&buffer, size * sizeof(value_type)));
+#endif // HYDROGEN_HAVE_CUB
+    return pointer(buffer);
```

Comment:
I think the typical cast notation (`(pointer) buffer` or a `static_cast`) is better.

Reply:
Thrust pointers are implemented in classes, not primitives (https://thrust.github.io/doc/classthrust_1_1device__ptr.html).